feat: Data table

cooper-joe opened the pull request:
This ones a big one! `Data table` component is probably going to be the most widely used component across DHIS2, so there's a lot to include here. I'm requesting review/eyes from you @dhis2/team-apps because I'd like some feedback on some of the included functionality. 

First, note that this component is intentionally complex in functionality. We have the `table` component if a simple table to display a few rows is all that is needed. Saying that, I'd like to discuss:

1. **Filtering**
I imagine the filtering functionality is probably the most complex. I've defined in the specs:

> By default the first 3 columns are displayed as filter buttons, this can be set to a custom value. Additional columns are available from the 'More' button. 

So, when using a data table a dev should be able to say 'Only display with 1 filter button, put the rest behind the 'More' button'. Is this possible?

Also, filtering is done by attribute type (text is filtered by text, number by min/max). However, if the column attribute was a data element groupset, ICD-10 diagnosis with 14,000 options, is this data table smart enough to say 'For this filter type, load all the options into a dropdown'. This seems...complex? What happens if there is some kind of filtering needed that we haven't anticipated, is it best to just load an empty popover and let the app do the actually filtering?

And! there are some discussions occurring [here](https://jira.dhis2.org/browse/DHIS2-6330) regarding how large data sets are loaded. I imagine this will have an effect on how filtering is carried out? 

2. **Row Actions**
The specs outline functionality for adding actions to table rows:

>Row actions are always displayed as the last column in a row. Primary row actions are displayed as text links in the row. A maximum of 2 primary row actions can be shown. Other secondary actions are available through the 'three-dots' overflow menu.
>
>A click action can be also be set per row that will be activated by clicking on that row. For example, clicking on a row may activate 'View' action and navigate to that row on a new page.

It makes sense for the row actions to be specced/discussed here in the design specs, but it seems to me that `data table row` should be it's own component. I just wanted to highlight that while the data table is one large design spec document, that doesn't need to be reflected in it's composition.

Will it be possible to create complex row actions if the `data table row` is a `ui-core` component? For example, an action might be 'when a row is clicked, if status=valid expand row, if status=invalid show tooltip with text...' etc. I guess my question is, how complex could a table row be while still being generic enough to be defined in `ui-core`?

3. **Inline editing**
Defined in the specs:

>If a cell is editable it will display an edit icon when the user hovers the cell/value.
>
>Clicking an editable cell opens a popover with the input controls for editing the value. The edit must be saved, at which point any relevant validations are run

I have a concrete use case for a question here. In the Capture app any inline edits will need to be checked against validation rules before being saved. Similar concerns to the filtering functionality included in `data table`. How will this be done? Is the 'inline edit' functionality defined in `data table` just opening a popover when an editable cell is clicked, or will we include saving/validation as well?

4.**Editable columns**
I discussed this [here](https://github.com/dhis2/ui-core/issues/386). In short, are there any problems with including this functionality in the `data table` component? (click button, open modal with dumb-table inside, edit and save column config). @Mohammer5 raised some concerns on the linked issue.

5.**Status badges**
From the design specs:
>Status badges are a useful component to include in data tables that need to display a status for each displayed row. The 'status badge' component provides consistent styles that can be combined with custom text and icons. Use status badges where each table row has a state from a few possible shared states.

...is this a pipe dream? It would be ideal to get some consistency across apps but I don't know if this is too difficult to achieve. The way I see this working: 
```
Load patients as rows in `data-table`
For each row: 
    if patient status=complete 
        status column = Neutral status badge with text 'Complete' icon 'check'.
    if patient status=overdue
        status column = Error status badge with text 'Overdue' icon 'clock'.
and so on...
```

`status badge` would need to be defined as a separate atom. My concern here is that adding this kind of logic inside a data table and inside in a table row somehow feels wrong to me? Advice appreciated!

---

My concerns here are mostly: is this component so complex that it will be a nightmare to build and implement? If so, are we better off defining separate tables (inline editable table, reorderable table...)? Or, do we leave out actual functionality and just build the wrappers where it should be included by the app itself?

All comments, suggestions and feedback on this component are very much appreciated.

ismay commented:
I talked to Joe on slack a bit about 1., as it wasn't clear to me what UI element what meant. Joe added a commit with an example, but for visibility, the filters are these kinds of controls:

![image](https://user-images.githubusercontent.com/7355199/63691215-e646ae80-c80e-11e9-9bcb-db80093718aa.png)

(So the `First name`, `Last name` and other buttons at the top. See commit e661900)

ismay commented:
Things that stand out to me (I might have missed parts of the discussion):

For 1. I guess the solution would be different for filtering local items vs retrieving filtered items from the backend. Are we building this table to do both types of filtering?

For 3., an edit icon that's only visible on hover won't be visible for users that aren't using a mouse. I'd say have it always visible if it's relevant to people that aren't using a mouse as well.

That's just what stood out to me. The component as a whole does seem quite big and complex. I don't know if I can comment on the feasibility of this as a reusable component.

Mohammer5 commented:
> By default the first 3 columns are displayed as filter buttons, this can be set to a custom value. Additional columns are available from the 'More' button.

I don't like this idea too much. I'd prefer if the app could define which columns should be display.
It makes sense to group columns by their relations, but that doesn't necessarily mean that these will be the ones being used to sort the table. I. e. it could make a lot more sense to sort by last name and creation date or update date, but the first 5 rows contain names and/or personal details.

> Also, filtering is done by attribute type (text is filtered by text, number by min/max). However, if the column attribute was a data element groupset, ICD-10 diagnosis with 14,000 options, is this data table smart enough to say 'For this filter type, load all the options into a dropdown'. This seems...complex?

I think this stuff should be provided by the backend. 
Users can have very slow computer (especially in SE-Asia and Africa), so there should be no heavy computation on the client ever (at least that's my opinion).

Which brings me to the next point:
We won't display all rows, most likely there will be pages. So naturally apps should only load the rows they are going to display instead of 20k+ possible rows.

**This means: Filtering and sorting will always happen on the backend** (as well as sorting, I haven't thought about that in dhis2/ui-core#386 ...) unless the app explicitly wants to do that otherwise. 
But it will never be done by the `DataTable` componen.

<hr />

> A maximum of 2 primary row actions can be shown

I bit nit-picky, but there shouldn't be two primary actions, at least in UX terms as that will confuse users.

> Will it be possible to create complex row actions if the data table row is a ui-core component? For example, an action might be 'when a row is clicked, if status=valid expand row, if status=invalid show tooltip with text...' etc. I guess my question is, how complex could a table row be while still being generic enough to be defined in ui-core?

Yes an no.. When taking the [current interpretation of what library is meant for what](https://github.com/dhis2/notes/issues/60), it should be possible to add this to `ui-core`. At the same time, components in `ui-core` should be as flexible as possible, and this is already quite opinionated, so I think it makes sense to keep this logic contained in the `DataTable` component.

In my opinion, the table row in `ui-core` should not have any logic, so that there are no constraints, you should be able to build anything that needs a table row, with any `onClick` logic.
(I think it will make more sense to put the `onClick` on the table cells that don't contain actions, otherwise users can click slightly next to an action button, but still inside the cell, which would trigger the `onClick` on the row)

<hr />

> Is the 'inline edit' functionality defined in data table just opening a popover when an editable cell is clicked, or will we include saving/validation as well?

I think it makes sense to have a full-fledged form in the component, which will handle validation.
This could be done with `@dhis2/ui-forms`.
I think it should be possible to define a `Field` for each row, which will contain validation as well, it could look similar to this:

```jsx
<DataTable
  inlineEditing={[
    () => <Field name="fname" component={InputField} />,
    undefined, // no editing of 2nd column possible,
    () => <Field name="mname" component={InputField} />,
  ]}
>
  ...
```

If the `DataTable` itself will use the api when submitting a valid value or if it simply calls a callback, is something that's still open.

<hr />

> 4.Editable columns

I think it will probably be a good idea to separate the `DataTable` and fetching the to-be-displayed data, that way we can display anything in the table, without having to implement exceptions/edge cases with might occur in a single app. Should be better to be able to define the exception in the app and to be able to maintain the special case if something changes without having to touch `ui-widgets`

<hr />

> My concern here is that adding this kind of logic inside a data table and inside in a table row somehow feels wrong to me? 

I assume these will be regular table cell content? Then I'd not put that into the `DataTable`

<hr />
<hr />

> is this component so complex that it will be a nightmare to build and implement?

I think, if we make this Component a side-effect free component (data, handling changes from inline-editing, sorting, filtering should all come/be done outside of this component), then this is still a complicated & complex component, but much simpler than if we want to include the functionality.

> are we better off defining separate tables (inline editable table, reorderable table...)

I can almost guarantee that someone will say "We need a re-orderable table with inline editing" at some point.. 

<hr />
<hr />

> For 3., an edit icon that's only visible on hover won't be visible for users that aren't using a mouse. I'd say have it always visible if it's relevant to people that aren't using a mouse as well.

100% agreed

cooper-joe commented:
Thanks for feedback on the points I raised here. I've modified the docs to address them and am merging this PR now to make it easilly available during build. Changes made just now:
- clarified the 'editable cell' section to say that editing controls within a popover will not be provided and must be provided by the app
- editable cells always display editable icon, not only on hover
- simplified filters. Filters will not automatically be available for _x_ number of columns, that was too confusing. Instead, the available filters must be defined when calling the component.

I have purposefully left out any mention of component composition here, but there is a discussion to be had about where the required components (e.g. TableRow, Pagination) should live and if they should be public/documented. We should have that discussion on https://github.com/dhis2/ui-widgets/pull/202.
